# **Welcome to Netflix Project**

In [2]:
# Import the required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [6]:
# Import the required dataset
df_ratings = pd.read_csv(r'D:\Intellipaat\Datasets\combined_data.txt', header=None, names=['Customer_id', 'Ratings'], usecols=[0, 1])
df_ratings

Customer_id  Ratings
0                 1:      NaN
1            1488844      3.0
2             822109      5.0
3             885013      4.0
4              30878      4.0
...              ...      ...
24058258     2591364      2.0
24058259     1791000      2.0
24058260      512536      5.0
24058261      988963      3.0
24058262     1704416      3.0

[24058263 rows x 2 columns]

Here, we see '1:' in start, this represent the 'Nth movie' and below the colon there are ratings given by customer for Nth movie. <br/>
means,

 -> below 1:, the ratings given by customers are for 1st movie <br/>
 -> below 2:, the ratings given by customers are for 2nd movie <br/>
 -> ..... till end

## Exploratory Data Analysis

In [7]:
# Display the total no. of movie for which customes rated
movie_count = df_ratings.isna().sum()[1]

Inference :- There are total 4499 movies which are rated by customers.

In [8]:
# Display the unique number of customers
customer_count = df_ratings['Customer_id'].nunique()
customer_count - movie_count            # No. of Movie also there, remove them.

470758

Inference :- There are total 470758 unique customers

In [9]:
# Store the movie number of each row in list 
movie_id = None
movie_no = []

# iterate over the dataset in customer id
for cust_id in df_ratings['Customer_id']:
    if ':' in cust_id:
        # update the current movie id
        movie_id = int(cust_id.replace(":", ""))
    movie_no.append(movie_id)

In [10]:
movie_no

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,


In [11]:
# Add new column "movie_id" in dataframe
df_ratings['Movie_id'] = movie_no

Now, This column contain the movie id of particular rows, and so nomore we need N: rows

In [12]:
df_ratings

Customer_id  Ratings  Movie_id
0                 1:      NaN         1
1            1488844      3.0         1
2             822109      5.0         1
3             885013      4.0         1
4              30878      4.0         1
...              ...      ...       ...
24058258     2591364      2.0      4499
24058259     1791000      2.0      4499
24058260      512536      5.0      4499
24058261      988963      3.0      4499
24058262     1704416      3.0      4499

[24058263 rows x 3 columns]

In [13]:
# Drop the N:/NaN records
df_ratings.dropna(inplace=True)

In [14]:
df_ratings

Customer_id  Ratings  Movie_id
1            1488844      3.0         1
2             822109      5.0         1
3             885013      4.0         1
4              30878      4.0         1
5             823519      3.0         1
...              ...      ...       ...
24058258     2591364      2.0      4499
24058259     1791000      2.0      4499
24058260      512536      5.0      4499
24058261      988963      3.0      4499
24058262     1704416      3.0      4499

[24053764 rows x 3 columns]

In [15]:
# Check the dtypes of df columns
df_ratings.dtypes

Customer_id     object
Ratings        float64
Movie_id         int64
dtype: object

In [16]:
# Convert the object dtype to numeric
df_ratings['Customer_id'] = df_ratings['Customer_id'].astype(int)

In [17]:
df_ratings.dtypes

Customer_id      int32
Ratings        float64
Movie_id         int64
dtype: object

**Pre- filtering** :- <br/>
- now we will remove all the users that have rated less number of movies
- also all those movies that has been rated less in number

In [18]:
# Create the df with respect to movie Id and their ratings count
movie_summmary = df_ratings.groupby('Movie_id')['Ratings'].agg(['count'])

In [19]:
movie_summmary

count
Movie_id       
1           547
2           145
3          2012
4           142
5          1140
...         ...
4495        614
4496       9519
4497        714
4498        269
4499        428

[4499 rows x 1 columns]

Here, <br/>
1 id is having 547, means 1st movie is rated by 547 customers <br/>
2 id is having 145, means 2nd movie is rated by 145 customers <br/>
.....

In [20]:
# Now set a benchmark for 60th percentile
movie_benchmark = round(movie_summmary['count'].quantile(0.6), 0)
movie_benchmark

908.0

In [21]:
# Create a list of movie index that need to be removed from the dataset
drop_movie_list = movie_summmary[movie_summmary['count'] < movie_benchmark].index
drop_movie_list

Index([   1,    2,    4,    7,    9,   10,   11,   12,   13,   14,
       ...
       4480, 4481, 4486, 4487, 4491, 4494, 4495, 4497, 4498, 4499],
      dtype='int64', name='Movie_id', length=2699)

In [22]:
# Display the total number of movie needs to be removed from dataset
len(drop_movie_list)

2699

In [23]:
# Create a df with respect to customers and total number of movies they've rated
customer_summary = df_ratings.groupby('Customer_id')["Ratings"].agg(["count"])

In [24]:
customer_summary

count
Customer_id       
6              153
7              195
8               21
10              49
25               4
...            ...
2649404         12
2649409         10
2649421          3
2649426         74
2649429         62

[470758 rows x 1 columns]

In [25]:
# set a benchmark for 60th percentile
customer_benchmark = round(customer_summary['count'].quantile(0.6), 0)
customer_benchmark

36.0

In [26]:
# Create a list of customer_id index that need to be removed from the dataset
drop_customer_list = customer_summary[customer_summary['count'] < customer_benchmark].index
drop_customer_list

Index([      8,      25,      33,      83,      94,     126,     130,     133,
           142,     149,
       ...
       2649337, 2649343, 2649351, 2649376, 2649379, 2649384, 2649401, 2649404,
       2649409, 2649421],
      dtype='int32', name='Customer_id', length=282042)

In [27]:
# Display the total number of customer needs to be removed from dataset
len(drop_customer_list)

282042

In [28]:
# Drop the cutomers & movies from the dataset
df_ratings = df_ratings[~df_ratings['Customer_id'].isin(drop_customer_list)]
df_ratings = df_ratings[~df_ratings['Movie_id'].isin(drop_movie_list)]

In [29]:
df_ratings

Customer_id  Ratings  Movie_id
696            712664      5.0         3
697           1331154      4.0         3
698           2632461      3.0         3
699             44937      5.0         3
700            656399      4.0         3
...               ...      ...       ...
24056842      1055714      5.0      4496
24056843      2643029      4.0      4496
24056844       267802      4.0      4496
24056845      1559566      3.0      4496
24056846       293198      3.0      4496

[19695836 rows x 3 columns]

## Model Building

In [30]:
# Load the other dataset having name of the movie
df_title = pd.read_csv(r'D:\Intellipaat\Datasets\Movie_titles.csv', encoding='ISO-8859-1', header=None, names=['Movie_id', 'Year', 'Name'], usecols=[0, 1, 2])

In [31]:
df_title

Movie_id    Year                                               Name
0             1  2003.0                                    Dinosaur Planet
1             2  2004.0                         Isle of Man TT 2004 Review
2             3  1997.0                                          Character
3             4  1994.0                       Paula Abdul's Get Up & Dance
4             5  2004.0                           The Rise and Fall of ECW
...         ...     ...                                                ...
17765     17766  2002.0  Where the Wild Things Are and Other Maurice Se...
17766     17767  2004.0                  Fidel Castro: American Experience
17767     17768  2000.0                                              Epoch
17768     17769  2003.0                                        The Company
17769     17770  2003.0                                       Alien Hunter

[17770 rows x 3 columns]

In [33]:
# This works only on google colab 
! pip install scikit-surprise

  Using cached scikit_surprise-1.1.4.tar.gz (154 kB)
  Installing build dependencies: started
  Installing build dependencies: still running...
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Failed to build scikit-surprise


  error: subprocess-exited-with-error
  
  × Building wheel for scikit-surprise (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [155 lines of output]
      C:\Users\vinit\AppData\Local\Temp\pip-build-env-yon8415u\overlay\Lib\site-packages\setuptools\config\_apply_pyprojecttoml.py:82: SetuptoolsDeprecationWarning: `project.license` as a TOML table is deprecated
      !!
      
              ********************************************************************************
              Please use a simple string containing a SPDX expression for `project.license`. You can also use `project.license-files`. (Both options available on setuptools>=77.0.0).
      
              By 2026-Feb-18, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ****************************************

  Using cached scikit_surprise-1.1.4.tar.gz (154 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Failed to build scikit-surprise


  error: subprocess-exited-with-error
  
  × Building wheel for scikit-surprise (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [155 lines of output]
      C:\Users\vinit\AppData\Local\Temp\pip-build-env-uu274nz_\overlay\Lib\site-packages\setuptools\config\_apply_pyprojecttoml.py:82: SetuptoolsDeprecationWarning: `project.license` as a TOML table is deprecated
      !!
      
              ********************************************************************************
              Please use a simple string containing a SPDX expression for `project.license`. You can also use `project.license-files`. (Both options available on setuptools>=77.0.0).
      
              By 2026-Feb-18, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ****************************************

Due to this we can't run the code file is not useful in jupyter notebook. <br/>
But it will work on google colab, import the file on google colab and run, we will get the output.

In [ ]:
# Import the libraries from surprise
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

In [ ]:
#  SVD (Singular value Decomposition) is used for recommendation engine etc

"""
The surprise library in Python is used for building and analyzing recommender systems. It provides tools to work with collaborative
 filtering algorithms,
which are commonly used to recommend items (such as movies, books, or products) to users based on their past interactions or preferences.
"""

In [ ]:
# Instance of reader class
reader = Reader()

In [ ]:
# We only take top 100k rows for quick runtime
data = Dataset.load_from_df(df_ratings[['Customer_id','Movie_id','Rating']][ : 100000], reader)

In [ ]:
data

In [ ]:
# Create the model
model = SVD()

SVD as it is a recomendation engione, the output of the model i.e., <br/> 
predicted data is the estimation score on which it will decide that movie to be recomended or Not. <br/>

The estimation score will be in the range of 1-5

In [ ]:
cross_validate(model, data, measures=["RMSE"], cv=3)

Run a cross validation procedure for a given algorithm, reporting accuracy measures and computation times. <br/><br/>

Here, cv = 3 means, the data is splited into 3 part randomly i.e., (1, 2, 3) and then <br/>
1. First time, 1 & 2 taken as training data and 3 will taken as testing data
2. second time 2 & 3 taken as training data and 1 will taken as testing data
3. third time, 1 & 3 taken as training data and 2 will taken as testing data <br/>
and perform the training & testing, and computes the metrics for particular part and shows it.

**Note** :- <br/>
This method makes sure that each sub-part data has been taken as training and testing once. And shows the performance of each data

## Recomendation

In [ ]:
df_ratings

In [29]:
df_title.head()

Movie_id    Year                          Name
0         1  2003.0               Dinosaur Planet
1         2  2004.0    Isle of Man TT 2004 Review
2         3  1997.0                     Character
3         4  1994.0  Paula Abdul's Get Up & Dance
4         5  2004.0      The Rise and Fall of ECW

In [ ]:
# we will be recomending a movies to a single user

user_1331154 = df_title.copy()  # making a copy of data set for this indivisual user

this data is having all the movies including those with less than 908 review which had to be dropped.

In [ ]:
# Drop those moivies with less than 908 reviews

user_1331154 = user_1331154[~user_1331154["Movie_Id"].isin(drop_movie_list)]
user_1331154

In [ ]:
# here in this code we are creating a new coulmns infront of each movie name on which we decide to recomend a movie

# Note: this is only for one user 1331154
user_1331154["Estimate_score"] = user_1331154["Movie_Id"].apply(lambda x:model.predict(1331154, x).est)
user_1331154

In [ ]:
# Display top 5 movies that should be recomended to the user
user_1331154.sort_values("Estimate_score", ascending=False).head()

END

---